### DISCUSSION TITLE
Using `outputFormat: "function-body"` with `@mdx-js/rollup`

### DISCUSSION THREAD

**Discussion by openscript** [mdx-js/mdx#2466] [Q&A] [answered]
I want to split up compiling and running. So I thought to configure `outputFormat: "function-body"` for  `@mdx-js/rollup`. When I import a .mdx file I receive the following error:

```
@timelet/web: Error [RollupError]: Return statement is not allowed here
@timelet/web:     at getRollupError (file:///workspaces/timelet/node_modules/.pnpm/rollup@4.14.0/node_modules/rollup/dist/es/shared/parseAst.js:379:41)
@timelet/web:     at convertProgram (file:///workspaces/timelet/node_modules/.pnpm/rollup@4.14.0/node_modules/rollup/dist/es/shared/parseAst.js:1043:26)
@timelet/web:     at parseAstAsync (file:///workspaces/timelet/node_modules/.pnpm/rollup@4.14.0/node_modules/rollup/dist/es/shared/parseAst.js:2189:12)
@timelet/web:     at ssrTransformScript (file:///workspaces/timelet/node_modules/.pnpm/vite@5.2.8_@types+node@20.12.5/node_modules/vite/dist/node/chunks/dep-whKeNLxG.js:54420:15)
@timelet/web:     at loadAndTransform (file:///workspaces/timelet/node_modules/.pnpm/vite@5.2.8_@types+node@20.12.5/node_modules/vite/dist/node/chunks/dep-whKeNLxG.js:53973:11)
@timelet/web:     at instantiateModule (file:///workspaces/timelet/node_modules/.pnpm/vite@5.2.8_@types+node@20.12.5/node_modules/vite/dist/node/chunks/dep-whKeNLxG.js:54967:10) {
```

As soon as I remove `outputFormat: "function-body"` this error disappears. Do have an idea why this is happening? I'm on version 3.0.1.

**Comment by wooorm** [comment]
Yeah that's not going to works rollup can only accept whole files. Not a function body.
Rollup is a bundler.

**Reply by openscript** [reply]
Thank you for the very quick feedback.
So the way to go is to ditch `@mdx-js/rollup` and set up another context, while using `compile` and `run`?

**Reply by wooorm** [reply]
Yes. Or both. Depends a bit on what you're doing. Have you seen https://mdxjs.com/guides/mdx-on-demand/?

**Reply by openscript** [reply]
Yes, I've read through that earlier. That's how I found `outputFormat: "function-body"`.

Well I have a hierarchy of .mdx files and want to render them using [vike.dev](https://vike.dev/). The .mdx files are external and not in the same project as I'm rendering the pages.

If I have both, what I actually need, because some pages are local .mdx files, I have the challenge to define the context (e.g. mapping the components to mdx) only once. Do you have any hints before I head of? :)

Your suggestions are very much appreciated!

**Reply by wooorm** [reply]
Not sure what Vike has to do with it. If you need that fine. `mdx` is just JS function calls. How to integrate the two has more to do with Vike than with MDX, probably.

> I have the challenge to define the context (e.g. mapping the components to mdx) only once.

I don’t understand, what is “context”? What are “mapped components”?

**Reply by openscript** [reply]
With context I understand a configuration of mdx and with mapped components basically [this](https://mdxjs.com/table-of-components/), but I think it's easy just to have the configuration at a single place.

**Reply by wooorm** [reply, marked as answer]
`components` is an object of tag names to functions. So you can have that somewhere in JavaScript like any other old object. And pass it in, in different places!
